src/hera/_utils/type_util.py

```diff
@@ -0,0 +1,93 @@
+"""Module that handles types and annotations."""
+
+from typing import Any, Iterable, Optional, Tuple, Type, TypeVar, Union, cast, overload
+
+from typing_extensions import TypeAlias
+
+try:
+    from types import UnionType  # type: ignore
+except ImportError:
+    UnionType = Union  # type: ignore
+
+try:
+    from typing import Annotated, get_args, get_origin  # type: ignore
+except ImportError:
+    from typing_extensions import Annotated, get_args, get_origin  # type: ignore
+
+
+def is_annotated(annotation: Any):
+    """Check annotation has Annotated type or not."""
+    return get_origin(annotation) is Annotated
+
+
+_Types: TypeAlias = Union[type, Tuple["_Types", ...]]
+
+
+def has_annotated_metadata(annotation: Any, type_: _Types) -> bool:
+    """If given annotation has metadata typed type_, return True."""
+    args = get_args(annotation)
+    for arg in args[1:]:
+        if isinstance(arg, type_):
+            return True
+    return False
+
+
+def consume_annotated_type(annotation: Any) -> type:
+    """If the given type is annotated, return unsubscripted version. If not return itself."""
+    if is_annotated(annotation):
+        return get_args(annotation)[0]
+    return annotation
+
+
+T = TypeVar("T")
+V = TypeVar("V")
+
+
+@overload
+def consume_annotated_metadata(_: Any, __: Type[T]) -> Optional[T]: ...
+
+
+@overload
+def consume_annotated_metadata(_: Any, __: Tuple[Type[T], Type[V]]) -> Optional[Union[T, V]]: ...
+
+
+# FIXME: Currently, mypy cannot guess following type hint properly: https://github.com/python/mypy/issues/17700
```

Comment:
Could be a use case of `TypeVarTuple` but I haven't looked into it much.

Reply:
I looked into it, but it doesn't seem to be used in here. anyway thanks for suggestion!